File: mysql/proxysql/python/proxysql_monitor/transport.py
from __future__ import annotations

import queue
import subprocess
import threading
import time
from collections import deque
from typing import Deque, List, Optional, TextIO, Tuple


class TransportError(RuntimeError):
    """Raised when the persistent MySQL client cannot complete a request."""


_EOF = object()


class PersistentMySQLSession:
    """Serialize framed SQL requests over one persistent MySQL CLI process."""

    def __init__(
        self,
        login_path: str,
        *,
        mysql_bin: str = "mysql",
        queue_size: int = 4096,
        stderr_lines: int = 100,
    ) -> None:
        self.login_path = login_path
        self.mysql_bin = mysql_bin
        self.queue_size = queue_size
        self.stderr_lines = stderr_lines
        self.process: Optional[subprocess.Popen[str]] = None
        self._sequence = 0
        self._closed = False
        self._execute_lock = threading.Lock()
        self._stdout_queue: "queue.Queue[object]" = queue.Queue(maxsize=queue_size)
        self._stderr: Deque[str] = deque(maxlen=stderr_lines)
        self._overflow = threading.Event()
        self._threads: List[threading.Thread] = []

    @property
    def recent_stderr(self) -> Tuple[str, ...]:
        """Return the bounded diagnostic tail from the current generation."""
        return tuple(self._stderr)

    def __enter__(self) -> "PersistentMySQLSession":
        self.start()
        return self

    def __exit__(self, *_args: object) -> None:
        self.close()

    def start(self) -> None:
        """Start the MySQL child if it is not already alive."""
        if self._closed:
            raise TransportError("MySQL session is closed.")
        if self.is_alive():
            return
        self._stdout_queue = queue.Queue(maxsize=self.queue_size)
        self._stderr = deque(maxlen=self.stderr_lines)
        self._overflow.clear()
        try:
            process = subprocess.Popen(
                [
                    self.mysql_bin,
                    f"--login-path={self.login_path}",
                    "--batch",
                    "--raw",
                    "--skip-column-names",
                    "--unbuffered",
                    "--force",
                ],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                shell=False,
                text=True,
                bufsize=1,
            )
        except OSError as exc:
            raise TransportError(f"Cannot start MySQL client: {exc}") from exc
        self.process = process
        assert process.stdout is not None
        assert process.stderr is not None
        stdout_queue = self._stdout_queue
        stderr_buffer = self._stderr
        overflow = self._overflow
        self._threads = [
            threading.Thread(
                target=self._read_stdout,
                args=(process.stdout, stdout_queue, overflow),
                name=f"pxmon-stdout-{process.pid}",
                daemon=True,
            ),
            threading.Thread(
                target=self._read_stderr,
                args=(process.stderr, stderr_buffer),
                name=f"pxmon-stderr-{process.pid}",
                daemon=True,
            ),
        ]
        for thread in self._threads:
            thread.start()

    def is_alive(self) -> bool:
        """Return whether the current child process is running."""
        return self.process is not None and self.process.poll() is None

    def execute(self, sql: str, timeout: float = 5.0) -> List[str]:
        """Execute one framed SQL payload and return its tab-separated rows."""
        if timeout <= 0:
            raise ValueError("timeout must be greater than zero")
        with self._execute_lock:
            self.start()
            process = self.process
            assert process is not None
            if process.stdin is None:
                raise TransportError("MySQL client stdin is unavailable.")
            self._sequence += 1
            begin = f"__PXMON_BEGIN_{self._sequence}__"
            end = f"__PXMON_END_{self._sequence}__"
            payload = f"SELECT '{begin}';\n{sql.rstrip()}\nSELECT '{end}';\n"
            try:
                process.stdin.write(payload)
                process.stdin.flush()
            except (BrokenPipeError, OSError, ValueError) as exc:
                raise TransportError(self._error_message("MySQL write failed", exc)) from exc

            deadline = time.monotonic() + timeout
            collecting = False
            rows: List[str] = []
            while True:
                if self._overflow.is_set():
                    raise TransportError("MySQL stdout queue overflow.")
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise TransportError(self._error_message("MySQL query timed out"))
                try:
                    item = self._stdout_queue.get(timeout=remaining)
                except queue.Empty as exc:
                    raise TransportError(self._error_message("MySQL query timed out")) from exc
                if item is _EOF:
                    raise TransportError(self._error_message("MySQL client exited"))
                line = str(item)
                if line == begin:
                    collecting = True
                    rows.clear()
                elif line == end and collecting:
                    return rows
                elif collecting:
                    rows.append(line)

    def execute_with_retry(self, sql: str, timeout: float = 5.0) -> List[str]:
        """Execute a request, reconnecting and retrying exactly once on failure."""
        try:
            return self.execute(sql, timeout)
        except TransportError:
            self.reconnect()
            return self.execute(sql, timeout)

    def reconnect(self) -> None:
        """Replace the current process generation."""
        if self._closed:
            raise TransportError("MySQL session is closed.")
        self._stop_process()
        self.start()

    def close(self) -> None:
        """Close and reap the child process; safe to call repeatedly."""
        if self._closed:
            return
        self._closed = True
        self._stop_process()

    def _stop_process(self) -> None:
        process = self.process
        self.process = None
        if process is None:
            return
        if process.stdin is not None:
            try:
                process.stdin.close()
            except OSError:
                pass
        if process.poll() is None:
            process.terminate()
            try:
                process.wait(timeout=1.0)
            except subprocess.TimeoutExpired:
                process.kill()
                process.wait(timeout=1.0)
        else:
            process.wait()
        for thread in self._threads:
            thread.join(timeout=1.0)
        self._threads.clear()
        for stream in (process.stdout, process.stderr):
            if stream is not None:
                stream.close()
        self._stdout_queue = queue.Queue(maxsize=self.queue_size)

    def _error_message(self, prefix: str, exc: Optional[BaseException] = None) -> str:
        parts = [prefix]
        if exc is not None:
            parts.append(str(exc))
        if self._stderr:
            parts.append(self._stderr[-1])
        return ": ".join(part for part in parts if part)

    @staticmethod
    def _read_stdout(
        stream: TextIO,
        output: "queue.Queue[object]",
        overflow: threading.Event,
    ) -> None:
        try:
            for raw_line in stream:
                try:
                    output.put_nowait(raw_line.rstrip("\r\n"))
                except queue.Full:
                    overflow.set()
                    return
        finally:
            try:
                output.put_nowait(_EOF)
            except queue.Full:
                overflow.set()

    @staticmethod
    def _read_stderr(stream: TextIO, output: Deque[str]) -> None:
        for raw_line in stream:
            output.append(raw_line.rstrip("\r\n"))

File: mysql/proxysql/go/internal/transport/transport.go
package transport

import (
	"bufio"
	"context"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
)

type Config struct {
	LoginPath   string
	MySQLBinary string
	ArgsPrefix  []string
	QueueSize   int
	StderrLines int
}

type lineResult struct {
	line string
	err  error
}

type Session struct {
	cfg Config

	mu       sync.Mutex
	execMu   sync.Mutex
	cmd      *exec.Cmd
	stdin    io.WriteCloser
	lines    chan lineResult
	done     chan struct{}
	overflow chan struct{}
	stopRead chan struct{}
	waitErr  error
	readers  sync.WaitGroup
	sequence uint64
	closed   bool
	stderr   *ringBuffer
}

func New(cfg Config) *Session {
	if cfg.MySQLBinary == "" {
		cfg.MySQLBinary = os.Getenv("MYSQL_BIN")
		if cfg.MySQLBinary == "" {
			cfg.MySQLBinary = "mysql"
		}
	}
	if cfg.QueueSize <= 0 {
		cfg.QueueSize = 4096
	}
	if cfg.StderrLines <= 0 {
		cfg.StderrLines = 100
	}
	return &Session{cfg: cfg, stderr: newRingBuffer(cfg.StderrLines)}
}

func (s *Session) commandArgs() []string {
	args := append([]string{}, s.cfg.ArgsPrefix...)
	return append(args,
		"--login-path="+s.cfg.LoginPath,
		"--batch",
		"--skip-column-names",
		"--unbuffered",
		"--force",
	)
}

func (s *Session) Start(ctx context.Context) error {
	s.execMu.Lock()
	defer s.execMu.Unlock()
	return s.start(ctx)
}

func (s *Session) start(ctx context.Context) error {
	if err := ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return errors.New("MySQL session is closed")
	}
	if s.cmd != nil {
		select {
		case <-s.done:
		default:
			s.mu.Unlock()
			return nil
		}
	}
	hadProcess := s.cmd != nil
	s.mu.Unlock()
	if hadProcess {
		_ = s.stopProcess()
	}

	command := exec.Command(s.cfg.MySQLBinary, s.commandArgs()...)
	stdin, err := command.StdinPipe()
	if err != nil {
		return fmt.Errorf("create MySQL stdin: %w", err)
	}
	stdout, err := command.StdoutPipe()
	if err != nil {
		_ = stdin.Close()
		return fmt.Errorf("create MySQL stdout: %w", err)
	}
	stderr, err := command.StderrPipe()
	if err != nil {
		_ = stdin.Close()
		return fmt.Errorf("create MySQL stderr: %w", err)
	}
	if err := command.Start(); err != nil {
		_ = stdin.Close()
		return fmt.Errorf("start MySQL client: %w", err)
	}

	s.mu.Lock()
	s.cmd = command
	s.stdin = stdin
	s.lines = make(chan lineResult, s.cfg.QueueSize)
	s.done = make(chan struct{})
	s.overflow = make(chan struct{})
	s.stopRead = make(chan struct{})
	s.waitErr = nil
	s.stderr = newRingBuffer(s.cfg.StderrLines)
	lines, done, overflow, stopRead := s.lines, s.done, s.overflow, s.stopRead

	s.readers.Add(2)
	go s.scanStdout(stdout, lines, overflow, stopRead)
	go s.scanStderr(stderr, stopRead)
	go func() {
		err := command.Wait()
		s.mu.Lock()
		s.waitErr = err
		s.mu.Unlock()
		close(done)
	}()
	s.mu.Unlock()
	return nil
}

func (s *Session) Alive() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.cmd == nil {
		return false
	}
	select {
	case <-s.done:
		return false
	default:
		return true
	}
}

func (s *Session) Execute(ctx context.Context, sql string, timeout time.Duration) ([]string, error) {
	s.execMu.Lock()
	defer s.execMu.Unlock()
	return s.execute(ctx, sql, timeout)
}

func (s *Session) execute(ctx context.Context, sql string, timeout time.Duration) ([]string, error) {
	if timeout <= 0 {
		return nil, errors.New("timeout must be greater than zero")
	}
	if err := s.start(ctx); err != nil {
		return nil, err
	}

	s.mu.Lock()
	s.sequence++
	sequence := s.sequence
	stdin, lines, done, overflow := s.stdin, s.lines, s.done, s.overflow
	s.mu.Unlock()
	begin := "__PXMON_BEGIN_" + strconv.FormatUint(sequence, 10) + "__"
	end := "__PXMON_END_" + strconv.FormatUint(sequence, 10) + "__"
	payload := fmt.Sprintf("SELECT '%s';\n%s\nSELECT '%s';\n", begin, strings.TrimSpace(sql), end)
	if _, err := io.WriteString(stdin, payload); err != nil {
		return nil, s.withStderr("MySQL write failed", err)
	}

	timer := time.NewTimer(timeout)
	defer timer.Stop()
	collecting := false
	rows := make([]string, 0)
	for {
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-timer.C:
			return nil, s.withStderr("MySQL query timed out", nil)
		case <-overflow:
			return nil, errors.New("MySQL stdout queue overflow")
		case <-done:
			s.mu.Lock()
			err := s.waitErr
			s.mu.Unlock()
			return nil, s.withStderr("MySQL client exited", err)
		case result := <-lines:
			if result.err != nil {
				return nil, s.withStderr("read MySQL stdout", result.err)
			}
			switch {
			case result.line == begin:
				collecting = true
				rows = rows[:0]
			case result.line == end && collecting:
				return rows, nil
			case collecting:
				rows = append(rows, result.line)
			}
		}
	}
}

func (s *Session) ExecuteWithRetry(ctx context.Context, sql string, timeout time.Duration) ([]string, error) {
	s.execMu.Lock()
	defer s.execMu.Unlock()
	rows, err := s.execute(ctx, sql, timeout)
	if err == nil {
		return rows, nil
	}
	if reconnectErr := s.stopProcess(); reconnectErr != nil {
		return nil, errors.Join(err, reconnectErr)
	}
	if startErr := s.start(ctx); startErr != nil {
		return nil, errors.Join(err, startErr)
	}
	return s.execute(ctx, sql, timeout)
}

func (s *Session) Reconnect(ctx context.Context) error {
	s.execMu.Lock()
	defer s.execMu.Unlock()
	if err := s.stopProcess(); err != nil {
		return err
	}
	return s.start(ctx)
}

func (s *Session) Stderr() []string {
	return s.stderr.Values()
}

func (s *Session) Close() error {
	s.execMu.Lock()
	defer s.execMu.Unlock()
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil
	}
	s.closed = true
	s.mu.Unlock()
	return s.stopProcess()
}

func (s *Session) stopProcess() error {
	s.mu.Lock()
	command, stdin, done, stopRead := s.cmd, s.stdin, s.done, s.stopRead
	s.cmd, s.stdin = nil, nil
	s.mu.Unlock()
	if command == nil {
		return nil
	}
	if stdin != nil {
		_ = stdin.Close()
	}
	if command.Process != nil {
		_ = command.Process.Signal(syscall.SIGTERM)
	}
	timer := time.NewTimer(time.Second)
	select {
	case <-done:
		timer.Stop()
	case <-timer.C:
		if command.Process != nil {
			_ = command.Process.Kill()
		}
		<-done
	}
	close(stopRead)
	s.readers.Wait()
	return nil
}

func (s *Session) scanStdout(
	reader io.Reader,
	lines chan<- lineResult,
	overflow chan struct{},
	stop <-chan struct{},
) {
	defer s.readers.Done()
	scanner := bufio.NewScanner(reader)
	buffer := make([]byte, 64*1024)
	scanner.Buffer(buffer, 4*1024*1024)
	var overflowOnce sync.Once
	for scanner.Scan() {
		select {
		case lines <- lineResult{line: strings.TrimSuffix(scanner.Text(), "\r")}:
		case <-stop:
			return
		default:
			overflowOnce.Do(func() { close(overflow) })
			return
		}
	}
	if err := scanner.Err(); err != nil {
		select {
		case lines <- lineResult{err: err}:
		case <-stop:
		default:
			overflowOnce.Do(func() { close(overflow) })
		}
	}
}

func (s *Session) scanStderr(reader io.Reader, stop <-chan struct{}) {
	defer s.readers.Done()
	scanner := bufio.NewScanner(reader)
	for scanner.Scan() {
		select {
		case <-stop:
			return
		default:
			s.stderr.Add(strings.TrimSuffix(scanner.Text(), "\r"))
		}
	}
}

func (s *Session) withStderr(prefix string, err error) error {
	message := prefix
	if err != nil {
		message += ": " + err.Error()
	}
	values := s.Stderr()
	if len(values) > 0 {
		message += ": " + values[len(values)-1]
	}
	return errors.New(message)
}

type ringBuffer struct {
	mu     sync.Mutex
	values []string
	next   int
	full   bool
}

func newRingBuffer(size int) *ringBuffer {
	return &ringBuffer{values: make([]string, size)}
}

func (r *ringBuffer) Add(value string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.values[r.next] = value
	r.next = (r.next + 1) % len(r.values)
	if r.next == 0 {
		r.full = true
	}
}

func (r *ringBuffer) Values() []string {
	r.mu.Lock()
	defer r.mu.Unlock()
	if !r.full {
		return append([]string(nil), r.values[:r.next]...)
	}
	values := append([]string(nil), r.values[r.next:]...)
	return append(values, r.values[:r.next]...)
}
